File: src/p1019/java/solution.java
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        ArrayList<Integer> result = new ArrayList<>();
        for (ListNode p = head; p != null; p = p.next) {
            result.add(p.val);
        }
        int[] stk = new int[result.size()];
        int top = 0;
        for (int i = 0; i < result.size(); i++) {
            while (top > 0 && result.get(stk[top - 1]) < result.get(i)) {
                result.set(stk[--top], result.get(i));
            }
            stk[top++] = i;
        }
        while (top > 0) result.set(stk[--top], 0);
        return result.stream().mapToInt(i -> i).toArray();
    }
}

File: src/p1019/golang/solution.go
func nextLargerNodes(head *ListNode) (result []int) {
	for p := head; p != nil; p = p.Next {
		result = append(result, p.Val)
	}
	stk, top := make([]int, len(result)), 0
	for i, val := range result {
		for top > 0 && result[stk[top-1]] < val {
			result[stk[top-1]], top = val, top-1
		}
		stk[top], top = i, top+1
	}
	for top > 0 {
		result[stk[top-1]], top = 0, top-1
	}
	return result
}
